extensions/azure/processors/PutAzureDataLakeStorage.cpp

```diff
@@ -0,0 +1,167 @@
+  if (!flow_file) {
+    context->yield();
+    return;
+  }
+
+  const auto params = buildUploadParameters(context, flow_file);
+  if (!params) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  PutAzureDataLakeStorage::ReadCallback callback(flow_file->getSize(), azure_data_lake_storage_, *params, logger_);
+  session->read(flow_file, &callback);
+  auto result = callback.getResult();
+  if (result.result_code == storage::UploadResultCode::FILE_ALREADY_EXISTS) {
+    gsl_Expects(conflict_resolution_strategy_ != FileExistsResolutionStrategy::REPLACE_FILE);
+    if (conflict_resolution_strategy_ == FileExistsResolutionStrategy::FAIL_FLOW) {
+      logger_->log_error("Failed to upload file '%s/%s' to filesystem '%s' on Azure Data Lake storage because file already exists",
+        params->directory_name, params->filename, params->file_system_name);
+      session->transfer(flow_file, Failure);
+      return;
+    } else if (conflict_resolution_strategy_ == FileExistsResolutionStrategy::IGNORE_REQUEST) {
+      logger_->log_debug("Upload of file '%s/%s' was ignored because it already exits in filesystem '%s' on Azure Data Lake Storage",
+        params->directory_name, params->filename, params->file_system_name);
+      session->transfer(flow_file, Success);
+      return;
+    }
+  } else if (result.result_code == storage::UploadResultCode::FAILURE) {
+    logger_->log_error("Failed to upload file '%s/%s' to filesystem '%s' on Azure Data Lake storage", params->directory_name, params->filename, params->file_system_name);
+    session->transfer(flow_file, Failure);
+  } else {
+    session->putAttribute(flow_file, "azure.filesystem", params->file_system_name);
+    session->putAttribute(flow_file, "azure.directory", params->directory_name);
+    session->putAttribute(flow_file, "azure.filename", params->filename);
+    session->putAttribute(flow_file, "azure.primaryUri", result.primary_uri);
+    session->putAttribute(flow_file, "azure.length", std::to_string(flow_file->getSize()));
+    logger_->log_debug("Successfully uploaded file '%s/%s' to filesystem '%s' on Azure Data Lake storage", params->directory_name, params->filename, params->file_system_name);
+    session->transfer(flow_file, Success);
+  }
+}
+
+PutAzureDataLakeStorage::ReadCallback::ReadCallback(
+      uint64_t flow_size, storage::AzureDataLakeStorage& azure_data_lake_storage, const storage::PutAzureDataLakeStorageParameters& params, std::shared_ptr<logging::Logger> logger)
+  : flow_size_(flow_size),
+    azure_data_lake_storage_(azure_data_lake_storage),
+    params_(params),
+    logger_(std::move(logger)) {
+}
+
+int64_t PutAzureDataLakeStorage::ReadCallback::process(const std::shared_ptr<io::BaseStream>& stream) {
+  std::vector<uint8_t> buffer;
+  int read_ret = stream->read(buffer, flow_size_);
+  if (io::isError(read_ret)) {
+    return -1;
+  }
+
+  result_ = azure_data_lake_storage_.uploadFile(params_, buffer.data(), flow_size_);
```

Comment:
here we might have read less than `flow_size_`

Reply:
I suppose it's due to the underlying buffer's size? What would be the most secure way here to read? If we read the data to a temporary buffer, can we know what's the appropriate buffer size for our case?

Reply:
Added check for read size in c569c2d49a1aacfc4bbf84f88da5a7588a203324